src/backend/distributed/connection/shared_connection_stats.c

```diff
@@ -0,0 +1,590 @@
+		memset(isNulls, false, sizeof(isNulls));
+
+		char *databaseName = get_database_name(connectionEntry->key.databaseOid);
+		if (databaseName == NULL)
+		{
+			/* database might have been dropped */
+			continue;
+		}
+
+		values[0] = PointerGetDatum(cstring_to_text(connectionEntry->key.hostname));
+		values[1] = Int32GetDatum(connectionEntry->key.port);
+		values[2] = PointerGetDatum(cstring_to_text(databaseName));
+		values[3] = Int32GetDatum(connectionEntry->connectionCount);
+
+		tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+	}
+
+	UnLockConnectionSharedMemory();
+}
+
+
+/*
+ * RemoveAllSharedConnectionEntriesForNode gets a nodename and nodeport, and removes
+ * the corresponding entry in the shared memory for the current database.
+ */
+void
+RemoveAllSharedConnectionEntriesForNode(char *hostname, int port)
+{
+	SharedConnStatsHashKey connKey;
+
+	connKey.databaseOid = MyDatabaseId;
+	connKey.port = port;
+	strlcpy(connKey.hostname, hostname, MAX_NODE_LENGTH);
+
+	/* we're reading all shared connections, prevent any changes */
+	LockConnectionSharedMemory(LW_EXCLUSIVE);
+
+	bool entryFound = false;
+	hash_search(SharedConnStatsHash, &connKey, HASH_REMOVE, &entryFound);
+
+	UnLockConnectionSharedMemory();
+}
+
+
+/*
+ * GetMaxSharedPoolSize is a wrapper around MaxSharedPoolSize which is controlled
+ * via a GUC.
+ *  "0" means adjust MaxSharedPoolSize automatically by using MaxConnections
+ * "-1" means do not apply connection throttling
+ * Anything else means use that number
+ */
+int
+GetMaxSharedPoolSize(void)
+{
+	if (MaxSharedPoolSize == 0)
+	{
+		return MaxConnections;
+	}
+
+	return MaxSharedPoolSize;
+}
+
+
+/*
+ * WaitOrErrorForSharedConnection tries to increment the shared connection
+ * counter for the given hostname/port and the current database in
+ * SharedConnStatsHash.
+ *
+ * The function implements a retry mechanism. If the function cannot increment
+ * the counter withing the specificed amount of the time, it throws an error.
+ */
+void
+WaitOrErrorForSharedConnection(const char *hostname, int port)
+{
+	while (!TryToIncrementSharedConnectionCounter(hostname, port))
+	{
+		WaitForSharedConnection();
+	}
+
+	ConditionVariableCancelSleep();
+}
+
+
+/*
+ * TryToIncrementSharedConnectionCounter tries to increment the shared
+ * connection counter for the given nodeId and the current database in
+ * SharedConnStatsHash.
+ *
+ * The function first checks whether the number of connections is less than
+ * citus.max_shared_pool_size. If so, the function increments the counter
+ * by one and returns true. Else, the function returns false.
+ */
+bool
+TryToIncrementSharedConnectionCounter(const char *hostname, int port)
+{
+	if (GetMaxSharedPoolSize() == -1)
+	{
+		/* connection throttling disabled */
+		return true;
+	}
+
+	bool counterIncremented = false;
+	SharedConnStatsHashKey connKey;
+
+	strlcpy(connKey.hostname, hostname, MAX_NODE_LENGTH);
+	if (strlen(hostname) > MAX_NODE_LENGTH)
```

Comment:
Probably not with the existing call stacks to get to this function. But, I wanted to keep this in case someday other callers use `TryToIncrementSharedConnectionCounter ` which is not coming through `multiConnection` APIs.